File: contracts.py
"""Stable message contracts shared by every runbook pipeline stage."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class Contract(BaseModel):
    """Base contract that tolerates additive fields but never renames them."""

    model_config = ConfigDict(extra="allow")


class TaskEnvelope(Contract):
    role: Literal["diagnostician", "remediator"]
    payload: dict[str, Any]
    trace_id: str


class Alert(Contract):
    alert_id: str
    fired_at: str
    severity: Literal["critical", "warning", "info"]
    service: str
    metric: str
    value: float
    threshold: float
    labels: dict[str, str] = Field(default_factory=dict)
    annotations: dict[str, str] = Field(default_factory=dict)


class Evidence(Contract):
    source: Literal["graph", "linkup", "past_incident", "alert"]
    ref: str
    detail: str


class Hypothesis(Contract):
    hypothesis_id: str
    type: Literal[
        "recent_deploy",
        "dependency_failure",
        "past_pattern",
        "external_event",
        "unknown",
    ]
    root_cause_description: str
    affected_entity: str
    confidence: float = Field(ge=0.0, le=1.0)
    evidence: list[Evidence] = Field(default_factory=list)
    reasoning: str


class HypothesisSet(Contract):
    alert_id: str
    hypotheses: list[Hypothesis]
    linkup_hits: int = Field(default=0, ge=0)
    graph_query_ms: int = Field(ge=0)
    generated_at: str
    guild_handoff_id: str | None = None


ActionType = Literal[
    "rollback",
    "restart",
    "scale",
    "flush_cache",
    "notify",
    "diagnostic",
    "none",
]
SafetyClass = Literal["safe", "standard", "destructive"]


class ActionProposal(Contract):
    proposal_id: str
    alert_id: str
    target_hypothesis_id: str
    action_type: ActionType
    action_target: str
    action_params: dict[str, Any] = Field(default_factory=dict)
    safety_class: SafetyClass
    remediator_confidence: float = Field(ge=0.0, le=1.0)
    runbook_source: str
    reasoning: str
    guild_task_id: str | None = None


class ReviewerDecision(Contract):
    proposal_id: str
    decision: Literal["approve", "reject", "modify"]
    modified_action: ActionProposal | None = None
    reviewer_note: str | None = None
    timestamp: str
    guild_task_id: str | None = None


class Resolution(Contract):
    incident_id: str
    alert_id: str
    final_action: ActionProposal | None = None
    outcome: Literal["verified", "partial", "no_effect", "rejected"]
    reviewer_decision: ReviewerDecision
    total_latency_ms: int = Field(ge=0)
    cost_usd: float = Field(ge=0.0)


class BridgeComplaintRequest(BaseModel):
    """Additive complaint input accepted by the Complaint Bridge."""

    model_config = ConfigDict(extra="forbid")

    text: str = Field(min_length=1)
    service: str = "payments-api"
    channel: str | None = None
    external_id: str | None = None
    graph_enabled: bool = True
    severity: Literal["critical", "warning", "info"] = "warning"


class LinearTicketPreview(Contract):
    connector: Literal["linear"] = "linear"
    dry_run: bool = True
    configured: bool
    operation: Literal["issue_create"] = "issue_create"
    payload: dict[str, Any]


class GitHubPullRequestPreview(Contract):
    connector: Literal["github"] = "github"
    dry_run: bool = True
    configured: bool
    operation: Literal["draft_pull_request"] = "draft_pull_request"
    repository: str | None = None
    branch: str
    base: str
    draft: bool = True
    direct_base_writes: bool = False
    merge: bool = False
    title: str
    body: str


class SlackReplyPreview(Contract):
    connector: Literal["slack"] = "slack"
    dry_run: bool = True
    configured: bool
    operation: Literal["reply_preview"] = "reply_preview"
    channel: str | None = None
    thread_ts: str | None = None
    text: str


class BridgeActionPreviews(Contract):
    linear: LinearTicketPreview
    github: GitHubPullRequestPreview
    slack_reply: SlackReplyPreview


class EvidenceBrief(Contract):
    alert_id: str
    summary: str
    complaint_text: str
    current_stage: Literal[
        "accepted",
        "diagnosed",
        "proposed",
        "reviewed",
        "resolved",
    ]
    leading_hypothesis: str | None = None
    evidence: list[Evidence] = Field(default_factory=list)
    proposed_action: str | None = None
    decision: str | None = None
    outcome: str | None = None
    assembled_at: str


class BridgeReport(Contract):
    alert_id: str
    alert: Alert
    complaint: BridgeComplaintRequest
    hypotheses: HypothesisSet | None = None
    proposal: ActionProposal | None = None
    decision: ReviewerDecision | None = None
    resolution: Resolution | None = None
    action_previews: BridgeActionPreviews
    evidence_brief: EvidenceBrief
    sponsor_boundaries: dict[str, str]
    guild_mode: str
    connector_dry_run: bool

File: web/src/contracts.ts
export type Tab = "dashboard" | "inbox" | "approvals" | "memory" | "settings";
export type LoadState = "idle" | "loading" | "ready" | "error";
export type SafetyClass = "safe" | "standard" | "destructive";
export type DecisionValue = "approve" | "reject" | "modify";

export interface BridgeHealth {
  status?: string;
  ok?: boolean;
  dry_run?: boolean;
  connector_dry_run?: boolean;
  connector_live_writes_enabled?: boolean;
  github_repo_allowed?: string | null;
  slack_connected?: boolean;
  slack_configured?: boolean;
  linear_configured?: boolean;
  github_configured?: boolean;
  guild_mode?: string;
  guild_qualifying?: boolean;
  sponsor_health?: {
    status?: string;
    guild_mode?: string;
    guild_qualifying?: boolean;
    [key: string]: unknown;
  };
  integrations?: Record<string, boolean | string | null>;
  [key: string]: unknown;
}

export interface SlackMessage {
  id: string;
  text: string;
  user?: string;
  user_name?: string;
  channel?: string;
  name?: string;
  channel_name?: string;
  timestamp?: string;
  ts?: string;
  permalink?: string;
  [key: string]: unknown;
}

export interface Alert {
  alert_id: string;
  fired_at: string;
  severity: "critical" | "warning" | "info";
  service: string;
  metric: string;
  value: number;
  threshold: number;
  labels?: Record<string, string>;
  annotations?: Record<string, string>;
}

export interface Evidence {
  source: "graph" | "linkup" | "past_incident" | "alert";
  ref: string;
  detail: string;
}

export interface Hypothesis {
  hypothesis_id: string;
  type: string;
  root_cause_description: string;
  affected_entity: string;
  confidence: number;
  evidence: Evidence[];
  reasoning: string;
}

export interface HypothesisSet {
  alert_id: string;
  hypotheses: Hypothesis[];
  linkup_hits: number;
  graph_query_ms: number;
  generated_at: string;
  guild_handoff_id?: string | null;
}

export interface ActionProposal {
  proposal_id: string;
  alert_id: string;
  target_hypothesis_id: string;
  action_type: string;
  action_target: string;
  action_params: Record<string, unknown>;
  safety_class: SafetyClass;
  remediator_confidence: number;
  runbook_source: string;
  reasoning: string;
  guild_task_id?: string | null;
}

export interface ReviewerDecision {
  proposal_id: string;
  decision: DecisionValue;
  modified_action?: ActionProposal | null;
  reviewer_note?: string | null;
  timestamp: string;
  guild_task_id?: string | null;
}

export interface ActionEvent {
  proposal_id: string;
  alert_id: string;
  action_type: string;
  action_target: string;
  status: string;
  simulated: boolean;
  timestamp: string;
  message?: string;
}

export interface OutcomeEvent {
  proposal_id: string;
  alert_id: string;
  status: string;
  outcome?: string;
  observed_value?: number | null;
  threshold?: number;
  simulated: boolean;
  timestamp: string;
}

export interface Resolution {
  incident_id: string;
  alert_id: string;
  final_action?: ActionProposal | null;
  outcome: "verified" | "partial" | "no_effect" | "rejected";
  reviewer_decision: ReviewerDecision;
  total_latency_ms: number;
  cost_usd: number;
}

export interface ActionPreview {
  provider: "linear" | "github" | "slack" | string;
  title?: string;
  body?: string;
  target?: string;
  status?: string;
  dry_run?: boolean;
  [key: string]: unknown;
}

export type ConnectorName = "linear" | "github" | "slack";
export type ConnectorActionStatus =
  | "pending"
  | "running"
  | "succeeded"
  | "failed"
  | "skipped"
  | "replayed";

export interface ConnectorAction {
  connector: ConnectorName;
  operation: string;
  status: ConnectorActionStatus;
  dry_run: boolean;
  idempotency_key: string;
  url?: string | null;
  identifier?: string | null;
  branch?: string | null;
  posted_at?: string | null;
  error?: string | null;
  alert_id?: string;
  proposal_id?: string;
}

export interface ConnectorWritesRollup {
  status?: string;
  total?: number;
  pending?: number;
  running?: number;
  succeeded?: number;
  failed?: number;
  skipped?: number;
  replayed?: number;
  partial?: boolean;
  dry_run?: boolean;
  [key: string]: unknown;
}

export interface ConnectorWritesReport extends ConnectorWritesRollup {
  rollup?: ConnectorWritesRollup;
  actions?: ConnectorAction[];
  proposal_id?: string;
  alert_id?: string;
  started_at?: string;
  completed_at?: string;
}

export interface BridgeReport {
  alert_id?: string;
  incident_id?: string;
  evidence_brief?: string | string[] | Record<string, unknown>;
  action_previews?: ActionPreview[] | Record<string, unknown>;
  linear_preview?: Record<string, unknown>;
  github_preview?: Record<string, unknown>;
  slack_preview?: Record<string, unknown>;
  connector_writes?: ConnectorWritesReport;
  [key: string]: unknown;
}

export type RunbookEventName =
  | "alert"
  | "hypotheses"
  | "proposal"
  | "decision"
  | "action"
  | "outcome"
  | "resolution"
  | "bridge_report"
  | "connector_action";

export type RunbookEventPayload =
  | Alert
  | HypothesisSet
  | ActionProposal
  | ReviewerDecision
  | ActionEvent
  | OutcomeEvent
  | Resolution
  | BridgeReport
  | ConnectorAction;

export interface TimelineEvent {
  id: string;
  type: RunbookEventName;
  at: string;
  payload: RunbookEventPayload;
}
